Fix #1346: change fixity during import

andreasabel opened the pull request:
Fix #1346: change fixity during import

andreasabel committed:
[ clean ] removed unused import of Agda.Syntax.Position

andreasabel committed:
[ #1346 refactor ] move Fixity type to Syntax.Common

Also: define some more Null instances

andreasabel committed:
[ refactor ] simpler definition of recomputeInScope

(while working on #1346)

andreasabel committed:
[ #1346 ] parse fixity annotation in renaming directive

Doesn't do anything yet.

andreasabel committed:
[ #1346 refactor ] simplify applyImportDirective by inlining some defs

JacquesCarette commented:
This new fixity annotation syntax upon renaming is very cool!  It should definitely be documented too...

andreasabel committed:
[ #3997 ] applyImportDirective: removed quadratic behavior

andreasabel committed:
[ #3997 ] applyImportDirectiveM: removed some sources of quadratic behavior

andreasabel committed:
[ cosmetics ] add separator comments [ci skip]

andreasabel committed:
[ #3997 ] applyImportDirectiveM: removed remaining quadratic behavior

The concrete-to-abstract translation of directives used of #1714 an
association list, which has been replaced by a Map now.

andreasabel committed:
[ opt ] Use (Set QName) in RecCheck

(inspired by #3997).

andreasabel committed:
[ opt ] Use Utils.List.duplicates instead of \\ and nub

(Crusade against nub inspired by #3997.)

andreasabel committed:
[ cosmetics ] Maybe for <=1 list

(Seen on the crusade against nub.)

andreasabel committed:
[ against nub ] use IntSet in ClausePostChecks

(While #3997.)

andreasabel committed:
[ clean ] Data.Either.partitionEithers retires Utils.Either.mapEither,

mapEither f = partitionEithers . map f

andreasabel committed:
[ Utils.Tuple ] /\ is &&&

andreasabel committed:
[ Utils.List ] implement editDistance using Data.Array

Without constant-time access to the memo structure,
we do not get the Θ(n*m) time complexity.

andreasabel committed:
[ Utils.List ] editDistance: full use of arrays

- avoids repeated calls to length (sums to quadratic for linked lists)
- makes List.tails obsolete
- really meets the complexity n*m now

andreasabel committed:
[ Utils.List ] cosmetical rewrites

andreasabel committed:
[ Utils.List ] partitionMaybe is a variant of partitionEithers

andreasabel committed:
[ Utils.List ] comments: complexity annotation, thematic sections

[ci skip]

andreasabel committed:
[ Utils.List ] reorder content to fit into thematic groups

andreasabel committed:
[ Utils.(Assoc)List ] repetitive lookups via List.hasElem and AssocList.apply

These lookup functions create a search tree from the list (Set or Map),
to speed up subsequent lookups.

Beneficial even for small lists (e.g. 5) with moderately many
lookups (e.g. 100).

andreasabel committed:
[ tiny opt ] data KindsOfNames to filter by KindOfName

Optimizes the silly test (`elem` [minBound..maxBound]).
Re #3997.

andreasabel committed:
[ cosmetics ] use filterMaybe, applyUnless; inlined mapScopeInfo

andreasabel committed:
[ fixed #1346 ] change fixity via renaming

The new fixity overwrites the nameFixity in the A.Name.

gallais commented on test/Succeed/Issue1346.agda:
```diff
@@ -0,0 +1,49 @@
+private
+  test₁ : List⁺ Bool
+  test₁ = true ∷ false ∷ []  -- mixing _∷_ of _×_ and List
+
```
This is so sneaky. I love it!

andreasabel committed:
[ new ] KillRange instance for Strict.Maybe

andreasabel committed:
[ reorg ] move useScope to TypeChecking.Monad.State

andreasabel committed:
[ haddock ] fix markup error in Utils.List

andreasabel committed:
[ while #1346 ] simplify computeFixitiesAndPolarities

However, getting lens stFixitiesAndPolarities right was not easy without
triggering #1829.

andreasabel committed:
[ #1346 ] warn about useless fixity in module renaming

andreasabel committed:
[ #1346 ] CHANGELOG and documentation in user manual

andreasabel commented on src/full/Agda/TypeChecking/Rules/Def.hs:
```diff
@@ -554,16 +556,17 @@ checkSystemCoverage _ _ t cs = __IMPOSSIBLE__
+
 data ClausesPostChecks = CPC
-    { cpcPartialSplits :: [Int]
+    { cpcPartialSplits :: IntSet
```
@Saizan : I hope this change is not harmful or annoying.

It is not related to the topic of this PR, I just got into a `nub`-killing frenzy.

andreasabel commented:
@JacquesCarette : I added the documentation to the CHANGELOG and the relevant section of the user manual.

andreasabel committed:
[ #1346 ] tests for highlighting of renaming infix

andreasabel commented:
Travis passed, except that the AllStdLib compiler test on ghc 8.0 ran out of time, as so often...
